File: backend/mcp-code-intelligence-kotlin/src/main/kotlin/com/codeintel/analyzers/entrypoints/PatternRegistry.kt
/**
 * KSA-162: Pattern Registry — Manages framework detection patterns.
 */
package com.codeintel.analyzers.entrypoints

class PatternRegistry {
    private val frameworks: Map<String, FrameworkPatterns> = mapOf(
        "fastapi" to FrameworkPatterns(
            language = "python",
            imports = listOf("fastapi", "from fastapi"),
            decorators = DecoratorPatterns(
                handler = listOf("app.get", "app.post", "app.put", "app.delete", "app.patch",
                    "router.get", "router.post", "router.put", "router.delete", "router.patch"),
                prefix = listOf("APIRouter(prefix="),
            ),
            authIndicators = listOf("Depends(get_current_user)", "Security(", "Depends(auth"),
        ),
        "express" to FrameworkPatterns(
            language = "typescript",
            imports = listOf("express", "from 'express'", "from \"express\""),
            callPatterns = CallPatterns(
                handler = listOf("app.get", "app.post", "app.put", "app.delete", "app.patch",
                    "router.get", "router.post", "router.put", "router.delete", "router.patch"),
                mount = listOf("app.use"),
            ),
            authIndicators = listOf("authenticate", "passport", "authMiddleware", "requireAuth"),
        ),
        "nestjs" to FrameworkPatterns(
            language = "typescript",
            imports = listOf("@nestjs/common", "@nestjs/core"),
            decorators = DecoratorPatterns(
                handler = listOf("Get", "Post", "Put", "Delete", "Patch", "Head", "Options", "All"),
                prefix = listOf("Controller"),
            ),
            authIndicators = listOf("UseGuards", "AuthGuard", "@Auth", "JwtAuthGuard"),
        ),
        "spring" to FrameworkPatterns(
            language = "java",
            imports = listOf("org.springframework.web", "org.springframework.boot"),
            decorators = DecoratorPatterns(
                handler = listOf("GetMapping", "PostMapping", "PutMapping", "DeleteMapping", "PatchMapping", "RequestMapping"),
                prefix = listOf("RequestMapping", "RestController"),
            ),
            authIndicators = listOf("@PreAuthorize", "@Secured", "@RolesAllowed"),
        ),
        "ktor" to FrameworkPatterns(
            language = "kotlin",
            imports = listOf("io.ktor.server", "io.ktor.routing"),
            callPatterns = CallPatterns(
                handler = listOf("get(", "post(", "put(", "delete(", "patch(", "route("),
                mount = listOf("routing {", "install("),
            ),
            authIndicators = listOf("authenticate(", "principal", "jwt {"),
        ),
        "gin" to FrameworkPatterns(
            language = "go",
            imports = listOf("github.com/gin-gonic/gin"),
            callPatterns = CallPatterns(
                handler = listOf("GET(", "POST(", "PUT(", "DELETE(", "PATCH(", "Handle("),
                mount = listOf("Group("),
            ),
            authIndicators = listOf("AuthMiddleware", "authRequired", "JWTAuth"),
        ),
    )

    private val mainPatterns: Map<String, Pair<String, String>> = mapOf(
        "python" to ("if __name__ == \"__main__\":" to "MAIN"),
        "typescript" to ("process.argv" to "MAIN"),
        "javascript" to ("process.argv" to "MAIN"),
        "java" to ("public static void main(String" to "MAIN"),
        "kotlin" to ("fun main(" to "MAIN"),
        "go" to ("func main()" to "MAIN"),
    )

    fun getFramework(name: String): FrameworkPatterns? = frameworks[name]

    fun getFrameworkNames(): List<String> = frameworks.keys.toList()

    fun getFrameworksForLanguage(language: String): List<Pair<String, FrameworkPatterns>> =
        frameworks.filter { it.value.language == language }.map { it.key to it.value }

    fun getMainPattern(language: String): Pair<String, String>? = mainPatterns[language]
}

File: backend/mcp-code-intelligence-python/src/mcp_code_intel/analyzers/entry_points/pattern_registry.py
"""KSA-162: Pattern Registry — Framework detection patterns."""
from __future__ import annotations
from . import FrameworkPatterns

DEFAULT_FRAMEWORKS: dict[str, FrameworkPatterns] = {
    "fastapi": FrameworkPatterns(
        language="python", imports=["fastapi", "from fastapi"],
        decorators={"handler": ["app.get", "app.post", "app.put", "app.delete", "app.patch",
                                "router.get", "router.post", "router.put", "router.delete", "router.patch"],
                    "prefix": ["APIRouter(prefix="]},
        auth_indicators=["Depends(get_current_user)", "Security(", "Depends(auth"],
    ),
    "express": FrameworkPatterns(
        language="typescript", imports=["express", "from 'express'", 'from "express"'],
        call_patterns={"handler": ["app.get", "app.post", "app.put", "app.delete", "app.patch",
                                   "router.get", "router.post", "router.put", "router.delete", "router.patch"],
                       "mount": ["app.use"]},
        auth_indicators=["authenticate", "passport", "authMiddleware", "requireAuth"],
    ),
    "nestjs": FrameworkPatterns(
        language="typescript", imports=["@nestjs/common", "@nestjs/core"],
        decorators={"handler": ["Get", "Post", "Put", "Delete", "Patch", "Head", "Options", "All"],
                    "prefix": ["Controller"]},
        auth_indicators=["UseGuards", "AuthGuard", "@Auth", "JwtAuthGuard"],
    ),
    "spring": FrameworkPatterns(
        language="java", imports=["org.springframework.web", "org.springframework.boot"],
        decorators={"handler": ["GetMapping", "PostMapping", "PutMapping", "DeleteMapping", "PatchMapping", "RequestMapping"],
                    "prefix": ["RequestMapping", "RestController"]},
        auth_indicators=["@PreAuthorize", "@Secured", "@RolesAllowed"],
    ),
    "ktor": FrameworkPatterns(
        language="kotlin", imports=["io.ktor.server", "io.ktor.routing"],
        call_patterns={"handler": ["get(", "post(", "put(", "delete(", "patch(", "route("],
                       "mount": ["routing {", "install("]},
        auth_indicators=["authenticate(", "principal", "jwt {"],
    ),
    "gin": FrameworkPatterns(
        language="go", imports=["github.com/gin-gonic/gin"],
        call_patterns={"handler": ["GET(", "POST(", "PUT(", "DELETE(", "PATCH(", "Handle("],
                       "mount": ["Group("]},
        auth_indicators=["AuthMiddleware", "authRequired", "JWTAuth"],
    ),
}

MAIN_PATTERNS: dict[str, tuple[str, str]] = {
    "python": ('if __name__ == "__main__":', "MAIN"),
    "typescript": ("process.argv", "MAIN"),
    "javascript": ("process.argv", "MAIN"),
    "java": ("public static void main(String", "MAIN"),
    "kotlin": ("fun main(", "MAIN"),
    "go": ("func main()", "MAIN"),
}


class PatternRegistry:
    def __init__(self) -> None:
        self._frameworks = DEFAULT_FRAMEWORKS
        self._main_patterns = MAIN_PATTERNS

    def get_framework(self, name: str) -> FrameworkPatterns | None:
        return self._frameworks.get(name)

    def get_frameworks_for_language(self, language: str) -> list[tuple[str, FrameworkPatterns]]:
        return [(n, p) for n, p in self._frameworks.items() if p.language == language]

    def get_main_pattern(self, language: str) -> tuple[str, str] | None:
        return self._main_patterns.get(language)
